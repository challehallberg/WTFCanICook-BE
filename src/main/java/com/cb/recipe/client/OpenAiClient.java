package com.cb.recipe.client;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@Component
public class OpenAiClient {
    private final WebClient client;

    public OpenAiClient(@Qualifier("openAiWebClient") final WebClient webClient) {
        this.client = webClient;
    }

    public String processImage(final String requestBody) {
        //TODO(CB): Add retryability
        return client.post()
                .uri("/chat/completions")
                .bodyValue(requestBody)
                .retrieve()
                .bodyToMono(String.class)
                .block();
    }
}
