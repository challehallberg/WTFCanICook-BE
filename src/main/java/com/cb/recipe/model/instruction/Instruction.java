package com.cb.recipe.model.instruction;

import lombok.Builder;
import lombok.Data;

import java.util.List;

@Data
@Builder
public class Instruction {
    // The name of this instruction set (can be empty)
    //@Nullable
    private String name;

    // The steps associated with this instruction set
    private List<InstructionStep> steps;
}
