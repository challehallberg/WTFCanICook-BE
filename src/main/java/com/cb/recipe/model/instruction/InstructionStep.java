package com.cb.recipe.model.instruction;

import lombok.Builder;
import lombok.Data;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Optional;

@Data
@Builder
public class InstructionStep {
    // The order number of this step
    private int number;

    // The textual description of the step
    private String step;

    // Optional: the duration for this step (if provided)
    @Nullable
    private StepLength length;

    // A list of equipment names used in this step
    @Nullable
    private List<Equipment> equipment;

    // A list of ingredient names used in this step
    @Nullable
    private List<String> ingredients;

    public Optional<StepLength> getOptionalLength() {
        return Optional.ofNullable(length);
    }

    // Optional getter for equipment
    public Optional<List<Equipment>> getOptionalEquipment() {
        return Optional.ofNullable(equipment);
    }

    // Optional getter for ingredients
    public Optional<List<String>> getOptionalIngredients() {
        return Optional.ofNullable(ingredients);
    }
}