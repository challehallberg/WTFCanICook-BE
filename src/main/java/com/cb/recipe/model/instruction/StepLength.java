package com.cb.recipe.model.instruction;

import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class StepLength {
    // The duration number (e.g., 15)
    private double number;

    // The duration unit (e.g., "minutes")
    private String unit;
}