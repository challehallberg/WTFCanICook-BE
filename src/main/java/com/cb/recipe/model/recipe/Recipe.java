package com.cb.recipe.model.recipe;

import lombok.Builder;
import lombok.Data;

import java.math.BigDecimal;
import java.util.List;

@Data
@Builder
public class Recipe {
    private int id;
    private String title;
    private String image;
    private String imageType;
    private int likes;

    private int missedIngredientCount;
    private List<Ingredient> missedIngredients;

    private BigDecimal usedIngredientCount;
    private List<Ingredient> usedIngredients;

    private int readyInMinutes;
    private boolean vegan;
    private boolean vegetarian;
    private boolean glutenFree;
    private int servings;
}
