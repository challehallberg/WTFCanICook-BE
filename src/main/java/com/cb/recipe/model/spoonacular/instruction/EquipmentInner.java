package com.cb.recipe.model.spoonacular.instruction;
import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class EquipmentInner {
    /**
     * The equipment id.
     */
    private int id;

    /**
     * The equipment image.
     */
    private String image;

    /**
     * The name of the equipment.
     */
    private String name;

    /**
     * Optional temperature information for the equipment.
     */
    private TemperatureInner temperature;
}
