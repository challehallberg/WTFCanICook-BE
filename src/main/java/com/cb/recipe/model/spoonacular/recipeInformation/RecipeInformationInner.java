package com.cb.recipe.model.spoonacular.recipeInformation;

import lombok.Builder;
import lombok.Data;

import java.util.List;

@Data
@Builder
public class RecipeInformationInner {
    private int id;
    private String title;
    private String image;
    private String imageType;
    private int servings;
    private int readyInMinutes;
    private int cookingMinutes;
    private int preparationMinutes;
    private String license;
    private String sourceName;
    private String sourceUrl;
    private String spoonacularSourceUrl;
    private double healthScore;
    private double spoonacularScore;
    private double pricePerServing;
    private List<AnalyzedInstructionInner> analyzedInstructions; // can be empty
    private boolean cheap;
    private String creditsText;
    private List<String> cuisines;
    private boolean dairyFree;
    private List<String> diets;
    private String gaps;
    private boolean glutenFree;
    private String instructions;
    private boolean ketogenic;
    private boolean lowFodmap;
    private List<String> occasions;
    private boolean sustainable;
    private boolean vegan;
    private boolean vegetarian;
    private boolean veryHealthy;
    private boolean veryPopular;
    private boolean whole30;
    private int weightWatcherSmartPoints;
    private List<String> dishTypes;
    private List<ExtendedIngredientInner> extendedIngredients;
    private String summary;
    private WinePairingInner winePairing;
}
