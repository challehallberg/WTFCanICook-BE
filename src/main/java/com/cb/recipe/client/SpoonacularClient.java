package com.cb.recipe.client;

import com.cb.recipe.model.spoonacular.instruction.InstructionInner;
import com.cb.recipe.model.spoonacular.recipe.RecipeInner;
import com.cb.recipe.model.spoonacular.recipeInformation.RecipeInformationInner;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;

@Component
public class SpoonacularClient {

    private final WebClient client;

    //TODO(CB): Fix retryability
    public SpoonacularClient(@Qualifier("spoonacularWebClient") final WebClient webClient) {
        this.client = webClient;
    }


    /**
     * Calls GET /recipes/findByIngredients endpoint with the provided parameters.
     *
     * @param ingredients  a comma-separated list of ingredients (e.g., "apples,flour,sugar")
     * @param number       the maximum number of recipes to return
     * @param ranking      whether to maximize used ingredients (1) or minimize missing ingredients (2)
     * @param ignorePantry whether to ignore typical pantry items (true/false)
     * @return a Mono containing a list of Recipe objects
     */
    public List<RecipeInner> getRecipesByIngredients(final String ingredients, final int number, final int ranking, final boolean ignorePantry) {
        return client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/recipes/findByIngredients")
                        .queryParam("ingredients", ingredients)
                        .queryParam("number", number)
                        .queryParam("ranking", ranking)
                        .queryParam("ignorePantry", ignorePantry)
                        .queryParam("limitLicense", true)
                        .build())
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<List<RecipeInner>>() {
                })
                .block();
    }

    /**
     * Retrieves bulk recipe information for the given recipe IDs.
     * <p>
     * This method calls the Spoonacular API endpoint "/recipes/informationBulk" using the provided query parameters.
     * It accepts a comma-separated string of recipe IDs and a flag indicating whether to include nutrition data.
     * The JSON response is deserialized into a list of {@link RecipeInformationInner} objects.
     * </p>
     *
     * @param ids              A comma-separated string of recipe IDs (e.g., "715538,716429").
     * @param includeNutrition A boolean flag to include nutrition data in the recipe information.
     * @return A list of {@link RecipeInformationInner} objects corresponding to the provided recipe IDs.
     */
    public List<RecipeInformationInner> getRecipeInformationBulk(final String ids, final boolean includeNutrition) {
        return client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/recipes/informationBulk")
                        .queryParam("ids", ids)
                        .queryParam("includeNutrition", includeNutrition)
                        .build())
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<List<RecipeInformationInner>>() {
                })
                .block();
    }

    /**
     * Retrieves analyzed instructions for a given recipe.
     * <p>
     * This method calls the Spoonacular API endpoint "/recipes/{id}/analyzedInstructions" with the specified recipe id
     * and step breakdown flag. The JSON response is deserialized into a list of {@link InstructionInner} objects.
     * </p>
     *
     * @param id            the recipe id (e.g., 324694)
     * @param stepBreakdown whether to break down the recipe steps even more
     * @return a list of {@link InstructionInner} objects containing the analyzed instructions for the recipe
     */
    public List<InstructionInner> getAnalyzedInstructions(final int id, final boolean stepBreakdown) {
        return client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/recipes/{id}/analyzedInstructions")
                        .queryParam("stepBreakdown", stepBreakdown)
                        .build(id))
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<List<InstructionInner>>() {
                })
                .block();
    }
}
