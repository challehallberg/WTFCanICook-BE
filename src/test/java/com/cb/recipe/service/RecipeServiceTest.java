package com.cb.recipe.service;

import com.cb.recipe.client.SpoonacularClient;
import com.cb.recipe.faker.instruction.InstructionFaker;
import com.cb.recipe.faker.recipe.RecipeFaker;
import com.cb.recipe.faker.recipe.RecipeInformationInnerFaker;
import com.cb.recipe.faker.recipe.RecipeInnerFaker;
import com.cb.recipe.model.instruction.Instruction;
import com.cb.recipe.model.recipe.Recipe;
import com.cb.recipe.model.spoonacular.recipeInformation.RecipeInformationInner;
import com.cb.recipe.util.UnitTest;
import com.github.benmanes.caffeine.cache.Cache;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;
import java.util.Random;

import static org.mockito.ArgumentMatchers.*;

@ExtendWith(MockitoExtension.class)
class RecipeServiceTest extends UnitTest {
    @Mock
    private SpoonacularClient client;

    @Mock
    private Cache<String, List<Recipe>> recipeCache;

    @Mock
    private Cache<Integer, List<Instruction>> instructionsCache;

    private RecipeService recipeService;

    @BeforeEach
    void setUp() {
        recipeService = new RecipeService(client, recipeCache, instructionsCache);
    }

    @Test
    void givenCachedIngredients_whenGetRecipes_thenExpectedCachedRecipes() {
        final String ingredients = "apple,banana";
        Mockito.when(recipeCache.getIfPresent(ingredients))
                .thenReturn(Arrays.asList(RecipeFaker.recipe().create()));

        final List<Recipe> result = recipeService.getRecipe(ingredients);

        snapshot(result);
    }

    @Test
    void givenValidIngredients_whenGetRecipes_thenExpectNonCachedRecipes() {
        final String ingredients = "apple,banana";
        Mockito.when(recipeCache.getIfPresent(ingredients))
                .thenReturn(null);

        Mockito.when(client.getRecipesByIngredients(anyString(), anyInt(), anyInt(), anyBoolean()))
                .thenReturn(Arrays.asList(RecipeInnerFaker.recipe().create()));

        final List<Recipe> result = recipeService.getRecipe(ingredients);

        snapshot(result);
    }

    @Test
    void givenValidRecipeIds_whenGetRecipeInformationBulk_thenExpectBulkOfRecipeInformation() {
        final String recipeIds = "1,2,3";

        Mockito.when(client.getRecipeInformationBulk(recipeIds, false))
                .thenReturn(Arrays.asList(RecipeInformationInnerFaker.recipeInformation(new Random(1)).create(),
                        RecipeInformationInnerFaker.recipeInformation(new Random(2)).create(),
                        RecipeInformationInnerFaker.recipeInformation(new Random(3)).create()));

        final List<RecipeInformationInner> result = recipeService.getRecipeInformationByIds(recipeIds);

        snapshot(result);
    }

    @Test
    void givenCachedRecipeId_whenGetAnalyzedRecipeInstructions_thenExpectCachedInstructions() {
        final int recipeId = 123;

        Mockito.when(instructionsCache.getIfPresent(recipeId))
                .thenReturn(Arrays.asList(InstructionFaker.instruction().create()));

        final List<Instruction> result = recipeService.getAnalyzedRecipeInstructions(recipeId);

        snapshot(result);
    }
    //TODO
    /*
    @Test
    void testGetAnalyzedRecipeInstructions_CacheMiss() {
        // Given: cache miss, so client call is made
        int recipeId = 1;
        Mockito.when(instructionsCache.getIfPresent(recipeId))
                .thenReturn(null);

        InstructionInner dummyInstructionInner = new InstructionInner();
        // Set any required fields on dummyInstructionInner if needed.
        List<InstructionInner> dummyInstructionInners = List.of(dummyInstructionInner);
        Mockito.when(client.getAnalyzedInstructions(recipeId, false))
                .thenReturn(Arrays.asList(InstructionInner));

        // Use static mocking for InstructionMapper.map
        List<Instruction> mappedInstructions = List.of(Instruction.builder().name("Mapped Instruction").build());
        try (MockedStatic<InstructionMapper> instructionMapperMock = mockStatic(InstructionMapper.class)) {
            instructionMapperMock.when(() -> InstructionMapper.map(dummyInstructionInners))
                    .thenReturn(mappedInstructions);

            // When
            List<Instruction> result = recipeService.getAnalyzedRecipeInstructions(recipeId);

            // Then
            assertEquals(mappedInstructions, result);
            verify(instructionsCache).put(recipeId, mappedInstructions);
        }
    }*/
}