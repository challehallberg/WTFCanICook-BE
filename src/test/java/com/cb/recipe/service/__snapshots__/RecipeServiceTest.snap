com.cb.recipe.service.RecipeServiceTest.givenCachedIngredients_whenGetRecipes_thenExpectedCachedRecipes=[
  [
    {
      "id": 985,
      "title": "Recipe 985",
      "image": "https://example.com/image985.jpg",
      "imageType": "jpg",
      "likes": 88,
      "missedIngredientCount": 2,
      "missedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        }
      ],
      "usedIngredientCount": 3,
      "usedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        },
        {
          "id": 108,
          "name": "Ingredient 108",
          "amount": 3.6403890715990754,
          "unit": "cup",
          "image": "https://example.com/ingredient108.jpg"
        }
      ],
      "readyInMinutes": 14,
      "vegan": false,
      "vegetarian": false,
      "glutenFree": true,
      "servings": 4
    }
  ]
]


com.cb.recipe.service.RecipeServiceTest.givenCachedRecipeId_whenGetAnalyzedRecipeInstructions_thenExpectCachedInstructions=[
  [
    {
      "name": "Instruction Set 88",
      "steps": [
        {
          "number": 1,
          "step": "Step description 4",
          "equipment": [ ],
          "ingredients": [ ]
        },
        {
          "number": 2,
          "step": "Step description 7",
          "length": {
            "number": 21.611449199578143,
            "unit": "minutes"
          },
          "equipment": [ ],
          "ingredients": [
            "Ingredient 64",
            "Ingredient 63"
          ]
        },
        {
          "number": 3,
          "step": "Step description 5",
          "length": {
            "number": 27.41756394766841,
            "unit": "minutes"
          },
          "equipment": [ ],
          "ingredients": [ ]
        },
        {
          "number": 4,
          "step": "Step description 3",
          "equipment": [ ],
          "ingredients": [ ]
        },
        {
          "number": 5,
          "step": "Step description 10",
          "length": {
            "number": 26.069426661179147,
            "unit": "minutes"
          },
          "equipment": [
            {
              "name": "Equipment 5",
              "temperature": {
                "number": 215.989270869342,
                "unit": "Fahrenheit"
              }
            }
          ],
          "ingredients": [ ]
        },
        {
          "number": 6,
          "step": "Step description 6",
          "equipment": [
            {
              "name": "Equipment 75",
              "temperature": {
                "number": 219.49479796024917,
                "unit": "Fahrenheit"
              }
            }
          ],
          "ingredients": [
            "Ingredient 38",
            "Ingredient 78"
          ]
        }
      ]
    }
  ]
]


com.cb.recipe.service.RecipeServiceTest.givenValidIngredients_whenGetRecipes_thenExpectNonCachedRecipes=[
  [
    {
      "id": 985,
      "title": "Recipe 985",
      "image": "https://example.com/recipe985.jpg",
      "imageType": "jpg",
      "likes": 88,
      "missedIngredientCount": 2,
      "missedIngredients": [
        {
          "id": 313,
          "name": "Ingredient 313",
          "amount": 1.3261184393533094,
          "unit": "unit",
          "image": "https://example.com/ingredient313.jpg"
        },
        {
          "id": 606,
          "name": "Ingredient 606",
          "amount": 7.3966566481449405,
          "unit": "unit",
          "image": "https://example.com/ingredient606.jpg"
        }
      ],
      "usedIngredientCount": 3,
      "usedIngredients": [
        {
          "id": 317,
          "name": "Ingredient 317",
          "amount": 9.458788499037189,
          "unit": "unit",
          "image": "https://example.com/ingredient317.jpg"
        },
        {
          "id": 592,
          "name": "Ingredient 592",
          "amount": 9.433739340063727,
          "unit": "unit",
          "image": "https://example.com/ingredient592.jpg"
        },
        {
          "id": 376,
          "name": "Ingredient 376",
          "amount": 4.127662262827992,
          "unit": "unit",
          "image": "https://example.com/ingredient376.jpg"
        }
      ],
      "readyInMinutes": 0,
      "vegan": false,
      "vegetarian": false,
      "glutenFree": false,
      "servings": 0
    }
  ]
]


com.cb.recipe.service.RecipeServiceTest.givenValidRecipeIds_whenGetRecipeInformationBulk_thenExpectBulkOfRecipeInformation=[
  [
    {
      "id": 8985,
      "servings": 5,
      "readyInMinutes": 48,
      "cookingMinutes": 0,
      "preparationMinutes": 0,
      "healthScore": 0.0,
      "spoonacularScore": 0.0,
      "pricePerServing": 0.0,
      "cheap": false,
      "dairyFree": false,
      "glutenFree": false,
      "ketogenic": false,
      "lowFodmap": false,
      "sustainable": false,
      "vegan": false,
      "vegetarian": false,
      "veryHealthy": false,
      "veryPopular": false,
      "whole30": false,
      "weightWatcherSmartPoints": 0
    },
    {
      "id": 6108,
      "servings": 1,
      "readyInMinutes": 32,
      "cookingMinutes": 0,
      "preparationMinutes": 0,
      "healthScore": 0.0,
      "spoonacularScore": 0.0,
      "pricePerServing": 0.0,
      "cheap": false,
      "dairyFree": false,
      "glutenFree": false,
      "ketogenic": false,
      "lowFodmap": false,
      "sustainable": false,
      "vegan": true,
      "vegetarian": false,
      "veryHealthy": false,
      "veryPopular": false,
      "whole30": false,
      "weightWatcherSmartPoints": 0
    },
    {
      "id": 3734,
      "servings": 3,
      "readyInMinutes": 20,
      "cookingMinutes": 0,
      "preparationMinutes": 0,
      "healthScore": 0.0,
      "spoonacularScore": 0.0,
      "pricePerServing": 0.0,
      "cheap": false,
      "dairyFree": false,
      "glutenFree": false,
      "ketogenic": false,
      "lowFodmap": false,
      "sustainable": false,
      "vegan": false,
      "vegetarian": true,
      "veryHealthy": false,
      "veryPopular": false,
      "whole30": false,
      "weightWatcherSmartPoints": 0
    }
  ]
]